### DISCUSSION TITLE
constructor on Layers

### DISCUSSION THREAD

**Discussion by ilan-gold** [visgl/deck.gl#5587] [Q&A] [answered]
Hello,

I am trying to add 
```javascript
  constructor(args) {
    super(args);
  }
```
to my layers for the purpose of auto-generating types for TypeScript.  It seems to work only when I change all of my layer calls in `renderLayers` from 
```javascript
return new Layer(this.props, {
      prop1: value,
      ...
})
```
to
```javascript
return new Layer({
      ...this.props,
      prop1: value,
      ....
})
```

I think it has something to do with https://github.com/visgl/deck.gl/blob/c53f30ee0e870e4d602172a3f9f96250d76f50c1/modules/core/src/lifecycle/component.js#L11-L29 but I am not sure.  Thanks!

**Comment by ilan-gold** [comment, marked as answer]
Figured it out - I had to change

```javsascript
  constructor(args) {
    super(args);
  }
```
to
```javascript
constructor(props) {
    // needed for TypeScript types that are generated from the JSDoc
    super(...arguments);
  }
```

Sorry, silly error!

**Comment by ibgreen** [comment]
@ilan-gold We are interested in adding typescript typings to deck.gl. We have already done so for the remaining vis.gl frameworks. 

If you are working on this perhaps we can collaborate?

**Reply by ilan-gold** [reply]
@ibgreen I could be interested in that.  I'm no expert on TypeScript though but what better way to learn!?  Our use case revolves around getting our JSDocs to generate TypeScript definitions for us without actually writing any TypeScript - that way if the docs are right, the definitions are too.  That also includes non-deck.gl JavaScript though.  Do you guys have a roadmap for it?  Or will it be a similar deal as with luma.gl, going module by module?

**Reply by ibgreen** [reply]
Currently the typescript-ification usually goes like this:
- add `tsconfig.json` to repo, and add `tsc --noEmit` to `yarn test lint`
- Fix all places where the existing JS does not pass typescript checking (without adding any types)
- Start adding `.d.ts` files to public APIs, and add a `types` field to package.json
- Incrementally add types to internal files.
